### PR TITLE
Fix outstanding problems in claiming and add SOCKS5 support.

### LINKED ISSUE #8090: ACLK Claiming should support Socks5 proxy
##### Feature idea summary
ACLK Claiming should support Socks5 proxy

##### Expected behavior
- [ ] Clarify with Cloud/SRE about testing endpoint for Claiming
- [ ] Update and test claiming script to support Socks5 proxy (pass parameters from config file to override env_var)

### LINKED ISSUE #8263: Fix known issues with claiming
#### Summary

As we work through testing we are collecting issues with claiming together so that we can track them before the release.

#### Known issues

* Timeout handling. Before I set the proxy I run the claiming command (the command line stalled for many seconds, probably reached the 3*5min timeout limit) but the script reported “Node was successfully claimed.” which was not the case.
* `hostname` is not available on some systems but `$HOSTNAME` is set.
* Needs a debug flag to see the contents of `tmpin` and `tmpout`.
* The integer expression problem if the return code is blank

### PR DESCRIPTION
**Pull request by amoss**
##### Summary
Closes #8090 
Closes #8263

Fix the problems in running the claiming script.

Merging from other PR (will integrete shortly).

some things in claiming script have to be fixed as well for this to be 100% (see #8263) - therefore this PR works and brings SOCKS5 proxy but there are error reporting issues that will be fixes in #8263.

If agent configured with SOCKS5 proxy [agent_cloud_link] key proxy it should pass that setting to curl when claiming script is called from within netdata agent with netdata -D -Wclaim.

##### Component Name
ACLK

##### Test Plan

* Claiming against the local cloud env with a valid url.
* Claiming against the local cloud env with an invalid url.
* Claiming against the local cloud env with incorrect port.
* Claiming against the local cloud env with nonexistent hostname.
* The `-verbose` flag shows the JSON and raw curl command.
* The '-insecure` flag works against a self-signed cert.
* Claiming against the cloud test platform connects and rejects an invalid token.
* Claiming against the cloud test platform connects and claims with a valid token.
* Claiming against the cloud test platform with an incorrect port.
* Claiming against the cloud test platform with an nonexistent hosthame.
* Claiming against the cloud test platform with the wrong proxy port -> confusing error message.
* Claiming against the cloud test platform with the wrong proxy hostname -> confusing error message.
* Claiming against the cloud test platform with a proxy ip on an unreachable network -> confusing error message.
* Claiming with `https_proxy` set in the environment.
* Claiming using the `-proxy=` option on the script.
* Claiming using `-W claim "-url=....`

Merging from other PR (will integrate shortly):

Make sure claiming script can be found by $PATH (e.g. if you test with NETDATA installed in custom dir)!

Test with local cloud test environment hidden behind socks5 proxy. (example: run agent on VM that doesn't have direct access to local cloud environment)

In config file of Netdata set options:

[agent_cloud_link]
    proxy = socks5://X.X.X.X:YYYY
and for other test
    proxy = none

When curl not present in system and wget present. Show error when trying to use SOCKS with wget (as wget doesn't have support).

### CODE REVIEW

**Comment by amoss** [comment]
@mfundul I didn't see your comments before I merged in @underhood 's commits and Github won't show them to me now. Could you dig them out and stick them in a fresh comment or send them to me on Slack?

**Comment by mfundul** [comment]
> 
> 
> @mfundul I didn't see your comments before I merged in @underhood 's commits and Github won't show them to me now. Could you dig them out and stick them in a fresh comment or send them to me on Slack?

I had only approved the commits. No comments.

File: claim/claim.c [hunk 1]
```diff
@@ -46,10 +48,29 @@ void claim_agent(char *claiming_arguments)
...
+    if(proxy_type == PROXY_TYPE_SOCKS5)
+        snprintf(proxy_flag, CLAIMING_PROXY_LENGTH, "-proxy=\"%s\"", proxy_str);
+
     snprintfz(command_buffer,
               CLAIMING_COMMAND_LENGTH,
-              "exec netdata-claim.sh -hostname=%s -id=%s -url=%s %s",
-              netdata_configured_hostname,
+              "exec netdata-claim.sh %s -hostname=%s -id=%s -url=%s %s",
+              proxy_flag,
+              cloud_base_url,
```
Comment (mfundul) [inline claim/claim.c]: You're replacing the machine hostname with `cloud_base_url` and keeping `registry.cloud_base_url` as is. Whas this the intended use?

File: claim/netdata-claim.sh.in [hunk 2]
```diff
@@ -101,36 +101,48 @@ ID="unknown"
...
-	shift 1
+        case $arg in
+                -token=*) TOKEN=${arg:7} ;;
+                -url=*) URL_BASE=${arg:5} ;;
+                -id=*) ID=${arg:4} ;;
+                -rooms=*) ROOMS=${arg:7} ;;
+                -hostname=*) HOSTNAME=${arg:10} ;;
+                -verbose) VERBOSE=1 ;;
+                -insecure) INSECURE=1 ;;
+                -proxy=socks*) PROXY=${arg:7} ;;
```
Comment (mfundul) [inline claim/netdata-claim.sh.in]: Does this mean we only support socks?

Comment (underhood) [reply claim/netdata-claim.sh.in]: yes, for now next PR will change this and bring support for others

File: claim/claim.c (hunk 1 above)
Comment (underhood) [reply claim/claim.c]: should be:
```
    snprintfz(command_buffer,
              CLAIMING_COMMAND_LENGTH,
              "exec netdata-claim.sh %s -hostname=%s -id=%s -url=%s %s",
              proxy_flag,
              netdata_configured_hostname,
              localhost->machine_guid,
              cloud_base_url,
              claiming_arguments);
```

Comment (amoss) [reply claim/claim.c]: I just need to check why I thought the test passed before I change the code.

Comment (amoss) [reply claim/claim.c]: It was the PATH problem that @underhood mentioned in the summary. Test was failing to fail in the right way. Now I can see this in the log:
```
2020-03-17 08:32:43: netdata INFO  : MAIN : Waiting for claiming command to finish.
Token: ****************
Base URL: https://localhost:8443/
Id: 99fa2382-49b3-11ea-b9ff-448a5bd059c8
Rooms: 
Hostname: https://localhost:8443/
```
But I can also see this earlier in the sequence:
```
2020-03-17 08:32:42: netdata ERROR : MAIN : User rl=https://dev.private.netdata.cloud:443 -token=unAchqvrEB9nqo2LKM6788NdIMC6d
aH_6bPwikKfox3rlKm7xJb4dZFmlEASEK2IM1ZrYmCbT58fiXhjm4QTKltkzVPNnOHiVa4tJEIfv0JxM8rlzWiKXoK1E0lKgcxlLVx0NiQ -rooms=37925172-8ab
4-446d-b8bb-18538a3162be -proxy=socks5h://127.0.0.1:11081 -verbose is not present.
2020-03-17 08:32:42: netdata ERROR : MAIN : Cannot become user 'rl=https://dev.private.netdata.cloud:443 -token=unAchqvrEB9nqo
2LKM6788NdIMC6daH_6bPwikKfox3rlKm7xJb4dZFmlEASEK2IM1ZrYmCbT58fiXhjm4QTKltkzVPNnOHiVa4tJEIfv0JxM8rlzWiKXoK1E0lKgcxlLVx0NiQ -roo
ms=37925172-8ab4-446d-b8bb-18538a3162be -proxy=socks5h://127.0.0.1:11081 -verbose'. Continuing as we are.
```

Comment (amoss) [reply claim/claim.c]: Tracing it through gdb I can see that the reason is that the configuration has been corrupted but it is not clear why yet:
```
[global]
	# run as user = rl=https://dev.private.netdata.cloud:443 -token=unAchqvrEB9nqo2LKM6788NdIMC6daH_6bPwikKfox3rlKm7xJb4dZFmlEASEK2IM1ZrYmCbT58fiXhjm4QTKltkzVPNnOHiVa4tJEIfv0JxM8rlzWiKXoK1E0lKgcxlLVx0NiQ -rooms=37925172-8ab4-446d-b8bb-18538a3162be -proxy=socks5h://127.0.0.1:11081 -verbose
```

**Pull request merged**